core/src/main/java/ai/timefold/solver/core/impl/move/director/MoveDirector.java

```diff
@@ -108,21 +108,56 @@ public final <Entity_, Value_> void changeVariable(PlanningVariableMetaModel<Sol
-            return moveValueInList(variableMetaModel, entity, destinationIndex, sourceIndex);
         }
         var variableDescriptor = extractVariableDescriptor(variableMetaModel);
-        var toIndex = destinationIndex + 1;
-        externalScoreDirector.beforeListVariableChanged(variableDescriptor, entity, sourceIndex, toIndex);
-        var variable = (List<Value_>) variableDescriptor.getValue(entity);
-        var value = variable.remove(sourceIndex);
-        variable.add(destinationIndex, value);
-        externalScoreDirector.afterListVariableChanged(variableDescriptor, entity, sourceIndex, toIndex);
-        return value;
+        var fromIndex = Math.min(sourceIndex, destinationIndex);
+        var toIndex = Math.max(sourceIndex, destinationIndex) + 1;
+
+        externalScoreDirector.beforeListVariableChanged(variableDescriptor, sourceEntity, fromIndex, toIndex);
+        Value_ element = (Value_) variableDescriptor.removeElement(sourceEntity, sourceIndex);
+        variableDescriptor.addElement(sourceEntity, destinationIndex, element);
+        externalScoreDirector.afterListVariableChanged(variableDescriptor, sourceEntity, fromIndex, toIndex);
+
+        return element;
+    }
+
+    @Override
+    public <Entity_, Value_> void swapValuesBetweenLists(
+            PlanningListVariableMetaModel<Solution_, Entity_, Value_> variableMetaModel, Entity_ leftEntity, int leftIndex,
+            Entity_ rightEntity, int rightIndex) {
+        if (leftEntity == rightEntity) {
+            swapValuesInList(variableMetaModel, leftEntity, leftIndex, rightIndex);
+        } else {
+            var variableDescriptor = extractVariableDescriptor(variableMetaModel);
+            externalScoreDirector.beforeListVariableChanged(variableDescriptor, leftEntity, leftIndex, leftIndex + 1);
+            externalScoreDirector.beforeListVariableChanged(variableDescriptor, rightEntity, rightIndex, rightIndex + 1);
+            var oldLeftElement = variableDescriptor.setElement(leftEntity, leftIndex,
+                    variableDescriptor.getElement(rightEntity, rightIndex));
+            variableDescriptor.setElement(rightEntity, rightIndex, oldLeftElement);
+            externalScoreDirector.afterListVariableChanged(variableDescriptor, leftEntity, leftIndex, leftIndex + 1);
+            externalScoreDirector.afterListVariableChanged(variableDescriptor, rightEntity, rightIndex, rightIndex + 1);
+        }
+    }
+
+    @Override
+    public <Entity_, Value_> void swapValuesInList(PlanningListVariableMetaModel<Solution_, Entity_, Value_> variableMetaModel,
+            Entity_ entity, int leftIndex, int rightIndex) {
+        if (leftIndex == rightIndex) {
+            // Do nothing.
+        } else if (leftIndex > rightIndex) { // Always start from the lower index.
+            swapValuesInList(variableMetaModel, entity, rightIndex, leftIndex);
+        } else {
+            var variableDescriptor = extractVariableDescriptor(variableMetaModel);
```

Comment:
I suggest moving the logic to the previous method and implementing the suggested changes to the contract definition. This will lead to a single method that remains non complex.